### PR TITLE
Feature/2466 1d integrator

### PR DESCRIPTION
**Pull request by bbbales2**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py src/test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary

This is the pull request to add integrate 1d to the Stan language. (thanks for doing all the heavy lifting @bob-carpenter)

#### How to Verify

```
./runTests.py src/test/unit
```

I'm not sure how to be more specific on the test models than that.

#### Documentation

I added docs to the functions reference (src/docs/functions/functions.tex) and the old user's guide (src/docs/users-guide/examples.tex). I haven't done docs for the new user's guide yet but I will once the content is finalized.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by bbbales2** [comment]
@bgoodri Yo could I get you to look at what I wrote up for the integrate 1d user's guide? The diff is here: https://github.com/stan-dev/stan/pull/2627/files#diff-474183abc2a400f1d8fd7e3f17646d9a and here is a link to the pdf version that should be easier to read: [users-guide.pdf](https://github.com/stan-dev/stan/files/2366199/users-guide.pdf) (pg. 183)

In particular, do you have a better example to use than the normal cdf thing I did? It'd be good if the first example is something that someone might reasonably do, but maybe that's not possible.

Once I get the content nailed down I'll move it to Rmarkdown and get all the bits sorted into the right places w/ @bob-carpenter.

File: src/docs/users-guide/examples.tex [hunk 1]
```diff
@@ -8931,3 +8929,178 @@ \subsection{Maximum Number of Steps}
 warmup.  With the non-stiff solver, this may result in jumping into a
 stiff region of the parameter space, which would require a very small
 step size and very many steps to satisfy even modest tolerances.
+
+
+\chapter{Computing 1D Integrals}
```
Comment (bgoodri) [inline src/docs/users-guide/examples.tex]: D -> Dimensional

File: src/docs/users-guide/examples.tex [hunk 2]
```diff
@@ -8931,3 +8929,178 @@ \subsection{Maximum Number of Steps}
...
+%
+
+\noindent
+This function is expected to return the value of the integrand evaluated at point \code{x}. The
+argument \code{xc} is used in definite integrals to avoid loss of precision near
+the limits of integration and is set to NaN when either limit is infinite
+(see \refsection{integral-precision} for details on how to use this). The argument
+\code{theta} is used to pass in arguments of the integral
+that are a function of the parameters in our model. The arguments \code{x\_r}
+and \code{x\_i} are likewise used to pass in arguments of the integral that are data.
```
Comment (bgoodri) [inline src/docs/users-guide/examples.tex]: real and integer data respectively

File: src/docs/users-guide/examples.tex [hunk 3]
```diff
@@ -8931,3 +8929,178 @@ \subsection{Maximum Number of Steps}
...
+the Stan calls \code{negative\_infinity()} and \code{positive\_infinity()}.
+
+If both limits are either \code{negative\_infinity()} or \code{positive\_infinity()}, the integral
+and its gradients are set to zero.
+
+\subsection{Data versus Parameters}
+%
+The arguments for the real data \code{x\_r} and the integer data \code{x\_i}
+must be expressions that only involve data or transformed data variables.
+\code{theta}, on the other hand, must only involve parameters.
```
Comment (bgoodri) [inline src/docs/users-guide/examples.tex]: really only parameters? No transformed parameters involving parameters and data?

File: src/docs/users-guide/examples.tex [hunk 4]
```diff
@@ -8931,3 +8929,178 @@ \subsection{Maximum Number of Steps}
...
+
+The endpoints of integration can be data or parameters (and interally the
+derivatives of the integral with respect to the endpoints are handled
+with the Leibniz integral rule).
+
+\section{Integrator convergence}
+%
+The integral is performed with the iterative 1D quadrature methods implemented in the Boost library. If the nth estimate of the integral is denoted $I_n$ and the nth estimate of the norm of the integral is denoted $|I|_n$, the iteration is terminated when
```
Comment (bgoodri) [inline src/docs/users-guide/examples.tex]: perhaps include link to Boost doc

File: src/stan/lang/grammars/semantic_actions_def.cpp [hunk 5]
```diff
@@ -1888,6 +1890,104 @@ namespace stan {
...
     validate_algebra_solver_control_f;
 
+    void validate_integrate_1d::operator()(
+            integrate_1d& fx, const variable_map& var_map,
+            bool& pass, std::ostream& error_msgs) const {
+      pass = true;
+
+      // (1) name of function to integrate
+      if (has_rng_lp_suffix(fx.function_name_)) {
+        error_msgs << "integrated function may not be an _rng or _lp function,"
```
Comment (bgoodri) [inline src/stan/lang/grammars/semantic_actions_def.cpp]: I get `_rng` functions but why can't we integrate a function ending in `_lp`?

**Comment by bgoodri** [comment]
Maybe a Stan version of the log of the BesselK function would be better than the truncated normal example? https://github.com/stan-dev/stan/wiki/Stan-Development-Meeting-Agenda/0ca4e1be9f7fc800658bfbd97331e800a4f50011

**Comment by bob-carpenter** [comment]
@mitzimorris said she could review the language bits if you want a second pair of eyes on it.

**Comment by bob-carpenter** [comment]
Also, this is going to need to be changed once I merge @mitzimorris's language update and she can also handle doing that on the branch.

**Comment by bbbales2** [comment]
@bgoodri Where'd you get the formula for this Bessel function? I tried coding up eq. 5 here:http://mathworld.wolfram.com/ModifiedBesselFunctionoftheSecondKind.html. The values seem to match but the way you wrote it always converges way faster.

And the main purpose of this would be for weird Matern kernels?

**Comment by bgoodri** [comment]
Equation 26 of
https://scholar.google.com/scholar?cluster=2908870453394922596&hl=en&as_sdt=5,33&sciodt=0,33

**Comment by bbbales2** [comment]
@bgoodri I bailed on the Bessel function example. It seemed like I was going to end up talking a lot about finicky numerics (how you split the integral in two to avoid overflow) and a model I don't really know anything about (a Matern GP with nu not one of the standard things) instead of just focusing on the integrator interface.

A very small amount of the LaTeX is in the git history (https://github.com/stan-dev/stan/pull/2627/commits/52c2590aa15c037c472a36eecbea8c27c9c8d4c6) if it was worth coming back to later once this is a real function that we can test in actual inferences.

**Comment by bbbales2** [comment]
@mitzimorris I think @bgoodri is correct here: https://github.com/stan-dev/stan/pull/2627#pullrequestreview-153819198 . We should be able to integrate _lpdf functions, I think. I'm not 100% sure what they are but if they're just functions that are pdfs, then we should definitely be able to integrate those.

**Comment by bgoodri** [comment]
Is there anything left to do here except fix the merge conflicts?

**Comment by ShaunMcDonald1021** [comment]
Apologies if this isn't quite the right place.

The current version of the 2.19 user guide on the Documentation page (https://mc-stan.org/docs/2_19/stan-users-guide-2_19.pdf) doesn't seem to have Ben's integrate_1d section. Is there a plan to add this?

**Comment by mitzimorris** [comment]
yes, absolutely!  looking into it

**Comment by seantalts** [comment]
Is there an issue that tracks what the behavior of this is? Trying to replicate in stanc3 and the diff is large.

**Pull request merged**